src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -481,6 +487,29 @@ private void AbortCore(Exception abortReason)
             _inputFlowControl.Abort();
         }
 
+        public void IncrementActiveStreamCount()
+        {
+            // If increment is called, it should only be called if we haven't decremented before.
+            // This method is only called once from HttpConnection.
+            Debug.Assert(!_needsDecrement);
+
+            _needsDecrement = true;
+            _context.StreamLifetimeHandler.IncrementActiveClientStreamCount();
+        }
+
+        public void DecrementActiveStreamCount()
```

Comment:
DecrementActiveStreamCount needs a lock, it can be called from an app thread or from the main connection loop. ApplyCompletionFlag already does something similar, re-use it's _completionLock?